// @kitware/vtk.js/Interaction/Widgets.js
import vtkAbstractWidget from './Widgets/AbstractWidget.js';
import vtkDistanceRepresentation from './Widgets/DistanceRepresentation.js';
import vtkDistanceWidget from './Widgets/DistanceWidget.js';
import vtkHandleRepresentation from './Widgets/HandleRepresentation.js';
import vtkHandleWidget from './Widgets/HandleWidget.js';
import vtkImageCroppingRegionsRepresentation from './Widgets/ImageCroppingRegionsRepresentation.js';
import vtkImageCroppingRegionsWidget from './Widgets/ImageCroppingRegionsWidget.js';
import vtkLabelRepresentation from './Widgets/LabelRepresentation.js';
import vtkLabelWidget from './Widgets/LabelWidget.js';
import vtkLineRepresentation from './Widgets/LineRepresentation.js';
import vtkLineWidget from './Widgets/LineWidget.js';
import vtkOrientationMarkerWidget from './Widgets/OrientationMarkerWidget.js';
import vtkPiecewiseGaussianWidget from './Widgets/PiecewiseGaussianWidget.js';
import vtkPointPlacer from './Widgets/PointPlacer.js';
import vtkResliceCursor from './Widgets/ResliceCursor/ResliceCursor.js';
import vtkResliceCursorLineRepresentation from './Widgets/ResliceCursor/ResliceCursorLineRepresentation.js';
import vtkResliceCursorWidget from './Widgets/ResliceCursor/ResliceCursorWidget.js';
import vtkSphereHandleRepresentation from './Widgets/SphereHandleRepresentation.js';
import vtkWidgetRepresentation from './Widgets/WidgetRepresentation.js';

var Widgets = {
  vtkAbstractWidget: vtkAbstractWidget,
  vtkDistanceRepresentation: vtkDistanceRepresentation,
  vtkDistanceWidget: vtkDistanceWidget,
  vtkHandleRepresentation: vtkHandleRepresentation,
  vtkHandleWidget: vtkHandleWidget,
  vtkImageCroppingRegionsRepresentation: vtkImageCroppingRegionsRepresentation,
  vtkImageCroppingRegionsWidget: vtkImageCroppingRegionsWidget,
  vtkLabelRepresentation: vtkLabelRepresentation,
  vtkLabelWidget: vtkLabelWidget,
  vtkLineRepresentation: vtkLineRepresentation,
  vtkLineWidget: vtkLineWidget,
  vtkOrientationMarkerWidget: vtkOrientationMarkerWidget,
  vtkPiecewiseGaussianWidget: vtkPiecewiseGaussianWidget,
  vtkPointPlacer: vtkPointPlacer,
  vtkResliceCursor: vtkResliceCursor,
  vtkResliceCursorLineRepresentation: vtkResliceCursorLineRepresentation,
  vtkResliceCursorWidget: vtkResliceCursorWidget,
  vtkSphereHandleRepresentation: vtkSphereHandleRepresentation,
  vtkWidgetRepresentation: vtkWidgetRepresentation
};

export { Widgets as default };
